datafusion/jit/src/compile.rs

```diff
@@ -0,0 +1,184 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Compile DataFusion Expr to JIT'd function.
+
+use datafusion_common::Result;
+
+use crate::api::Assembler;
+use crate::{
+    api::GeneratedFunction,
+    ast::{Expr as JITExpr, I64, PTR_SIZE},
+};
+
+/// Wrap JIT Expr to array compute function.
+pub fn build_calc_fn(
+    assembler: &Assembler,
+    jit_expr: JITExpr,
+    input_names: Vec<String>,
+) -> Result<GeneratedFunction> {
+    let mut builder = assembler.new_func_builder("calc_fn");
+    for input in &input_names {
+        builder = builder.param(format!("{}_array", input), I64);
+    }
+    let mut builder = builder.param("result", I64).param("len", I64);
+
+    let mut fn_body = builder.enter_block();
+    fn_body.declare_as("index", fn_body.lit_i(0))?;
+    fn_body.while_block(
+        |cond| cond.lt(cond.id("index")?, cond.id("len")?),
```

Comment:
Hmm, do we need sanity check for equal array lengths?

Reply:
The generated code is working on pointer directly, it might be hard to do these check inside it. I check inputs' length before pass them to generated code at [here](https://github.com/apache/arrow-datafusion/pull/2587/files#diff-4d319f92665cd672813b17b0693fad4817830e8cd19c719ec738d28fceee642aR92-R96). 
```rust
        if lhs.len() != rhs.len() {
            return Err(DataFusionError::NotImplemented(
                "Computing on different length arrays not yet supported".to_string(),
            ));
        }
```
But I agree that we should consider how to improve safety of generated code when its logic get complicated.